File: workers/research/src/black_book_research/adapters/federal/shared/retention.py
"""Candidate retention rules for federal archive adapters (BB-046)."""

from __future__ import annotations

from .types import FederalRejectedRecord, FederalRetentionRules


def _read_string(raw: dict[str, object], field: str) -> str | None:
    value = raw.get(field)
    if isinstance(value, str) and value.strip():
        return value.strip()
    return None


def qualifies_for_candidate_retention(
    raw: dict[str, object],
    rules: FederalRetentionRules,
) -> tuple[bool, str | None]:
    for field in rules.required_fields:
        if not _read_string(raw, field):
            return False, f"missing_required_field:{field}"

    title = _read_string(raw, "title")
    if not title or len(title) < rules.min_title_length:
        return False, "title_too_short"

    classification = _read_string(raw, "classification")
    if classification and classification not in rules.allowed_classifications:
        return False, "classification_not_allowed"

    if rules.require_canonical_url and not _read_string(raw, "canonicalUrl"):
        return False, "missing_canonical_url"

    return True, None


def partition_by_retention(
    records: list[dict[str, object]],
    rules: FederalRetentionRules,
) -> tuple[tuple[dict[str, object], ...], tuple[FederalRejectedRecord, ...]]:
    qualified: list[dict[str, object]] = []
    rejected: list[FederalRejectedRecord] = []

    for record in records:
        stable_identifier = (
            _read_string(record, "stableIdentifier")
            or _read_string(record, "id")
            or "unknown"
        )
        ok, reason = qualifies_for_candidate_retention(record, rules)
        if ok:
            qualified.append(record)
        else:
            rejected.append(
                FederalRejectedRecord(
                    stable_identifier=stable_identifier,
                    reason=reason or "rejected",
                )
            )

    return tuple(qualified), tuple(rejected)

File: packages/domain/src/adapters/federal/shared/retention.ts
/**
 * Candidate retention rules for federal archive adapters.
 * Only qualifying metadata and evidence fields proceed toward discovery.
 */
import type { FederalRejectedRecord, FederalRetentionRules, RawFederalExportRecord } from './types.js';

function readString(raw: RawFederalExportRecord, field: string): string | undefined {
  const value = raw[field];
  return typeof value === 'string' && value.trim() ? value.trim() : undefined;
}

export function qualifiesForCandidateRetention(
  raw: RawFederalExportRecord,
  rules: FederalRetentionRules,
): { qualified: true } | { qualified: false; reason: string } {
  for (const field of rules.requiredFields) {
    if (!readString(raw, field)) {
      return { qualified: false, reason: `missing_required_field:${field}` };
    }
  }

  const title = readString(raw, 'title');
  if (!title || title.length < rules.minTitleLength) {
    return { qualified: false, reason: 'title_too_short' };
  }

  const classification = readString(raw, 'classification');
  if (classification && !rules.allowedClassifications.includes(classification)) {
    return { qualified: false, reason: 'classification_not_allowed' };
  }

  if (rules.requireCanonicalUrl && !readString(raw, 'canonicalUrl')) {
    return { qualified: false, reason: 'missing_canonical_url' };
  }

  return { qualified: true };
}

export function partitionByRetention(
  records: readonly RawFederalExportRecord[],
  rules: FederalRetentionRules,
): {
  readonly qualified: readonly RawFederalExportRecord[];
  readonly rejected: readonly FederalRejectedRecord[];
} {
  const qualified: RawFederalExportRecord[] = [];
  const rejected: FederalRejectedRecord[] = [];

  for (const record of records) {
    const stableIdentifier =
      readString(record, 'stableIdentifier') ??
      readString(record, 'id') ??
      'unknown';
    const gate = qualifiesForCandidateRetention(record, rules);
    if (gate.qualified) {
      qualified.push(record);
    } else {
      rejected.push({ stableIdentifier, reason: gate.reason });
    }
  }

  return { qualified, rejected };
}
